hail/src/main/scala/is/hail/expr/types/encoded/EStruct.scala

```diff
@@ -0,0 +1,188 @@
+
+  def _decodedPType(requestedType: Type): PType = requestedType match {
+    case t: TInterval =>
+      val repr = t.representation
+      val pointType = _decodedPType(repr).asInstanceOf[PStruct].fieldType("start")
+      PInterval(pointType, required)
+    case t: TLocus => PLocus(t.rg, required)
+    case t: TStruct =>
+      val pFields = t.fields.map { case Field(name, typ, idx) =>
+        val pt = fieldType(name).decodedPType(typ)
+        PField(name, pt, idx)
+      }
+      PStruct(pFields, required)
+  }
+
+  override def _buildEncoder(pt: PType, mb: MethodBuilder, v: Code[_], out: Code[OutputBuffer]): Code[Unit] = {
+    val ft = pt.asInstanceOf[PStruct]
+    val addr = coerce[Long](v)
+    val writeMissingBytes = if (ft.size == size) {
+      out.writeBytes(addr, ft.nMissingBytes)
+    } else {
+      var c: Code[Unit] = Code._empty[Unit]
+      var j = 0
+      var n = 0
+      while (j < size) {
+        var b = const(0)
+        var k = 0
+        while (k < 8 && j < size) {
+          val f = fields(j)
+          if (!f.typ.required) {
+            val i = ft.fieldIdx(f.name)
+            b = b | (ft.isFieldMissing(addr, i).toI << k)
+            k += 1
+          }
+          j += 1
+        }
+        if (k > 0) {
+          c = Code(c, out.writeByte(b.toB))
+          n += 1
+        }
+      }
+      assert(n == nMissingBytes)
+      c
+    }
+
+    val writeFields = Code(fields.map { ef =>
+      val i = ft.fieldIdx(ef.name)
+      val pf = ft.fields(i)
+      val encodeField = ef.typ.buildEncoder(pf.typ, mb)
+      val v = Region.loadIRIntermediate(pf.typ)(ft.fieldOffset(addr, i))
+      ft.isFieldDefined(addr, i).mux(
+        encodeField(v, out),
+        Code._empty[Unit]
+      )
+    }: _*)
+
+    Code(writeMissingBytes, writeFields, Code._empty[Unit])
+  }
+
+  override def _buildInplaceDecoder(
+    pt: PType,
+    mb: MethodBuilder,
+    region: Code[Region],
+    addr: Code[Long],
+    in: Code[InputBuffer]
+  ): Code[Unit] = {
+    val mbytes = mb.newLocal[Long]("mbytes")
+
+    val t = pt.asInstanceOf[PStruct]
+    val readFields = fields.map { f =>
+      if (t.hasField(f.name)) {
+        val rf = t.field(f.name)
+        val readElemF = f.typ.buildInplaceDecoder(rf.typ, mb)
+        val rFieldAddr = t.fieldOffset(addr, rf.index)
+        if (f.typ.required)
+          readElemF(region, rFieldAddr, in)
+        else
+          Region.loadBit(mbytes, const(missingIdx(f.index).toLong)).mux(
+            t.setFieldMissing(addr, rf.index),
+            Code(
+              t.setFieldPresent(addr, rf.index),
+              readElemF(region, rFieldAddr, in)))
+      } else {
+        val skip = f.typ.buildSkip(mb)
+        if (f.typ.required)
+          skip(region, in)
+        else
+          Region.loadBit(mbytes, const(missingIdx(f.index).toLong)).mux(
+            Code._empty[Unit],
+            skip(region, in))
+      }
+    }
+
+    Code(
+      mbytes := region.allocate(const(1), const(nMissingBytes)),
```

Comment:
what's this allocation for? Can't you copy these bytes directly into `addr`? I know this won't work with all ptypes, but it will be faster in the common case (and we can restrict compatibility so that we know it always works)